CI: Run PHPUnit for core and some extensions

vedmaka opened the pull request:
Add a new workflow job, `test`, that will start a docker wiki (similar to the one used for the end-to-end testing) and run phpunit tests. The tests for MediaWiki core and a few extensions are run, using the configuration in `e2e/LocalSettings.php` - to add additional extensions to those tested in CI, just enable them there.

Significant work was needed to get various tests to pass when running in Taqasta, including patching test files to change their test configuration and skipping or deleting some tests. Ideally, all tests would pass without any of these modifications. Not all failures have been analyzed for a root cause yet.

Some of the fixes have already been applied to MediaWiki core and just are not included in MediaWiki 1.43.8 - fixes already applied to MediaWiki core are documented with links to the appropriate patch, and should be removed once Taqasta's version of MediaWiki incorporates those changes.

The fixes and removals are handled in the same `e2e/LocalSettings.php` file that controls the settings used for the tests.

WIK-471

DanielWTQ commented:
Given that this branch was 254 commits behind, rebase was a bit messy and probably lost some details

DanielButAtWork committed:
CI: Run PHPUnit for core and some extensions

Add a new workflow job, `test`, that will start a docker wiki (similar to the
one used for the end-to-end testing) and run phpunit tests. The tests for
MediaWiki core and a few extensions are run, using the configuration in
`e2e/LocalSettings.php` - to add additional extensions to those tested in CI,
just enable them there.

Significant work was needed to get various tests to pass when running in
Taqasta, including patching test files to change their test configuration and
skipping or deleting some tests. Ideally, all tests would pass without any of
these modifications. Not all failures have been analyzed for a root cause yet.

Some of the fixes have already been applied to MediaWiki core and just are not
included in MediaWiki 1.43.8 - fixes already applied to MediaWiki core are
documented with links to the appropriate patch, and should be removed once
Taqasta's version of MediaWiki incorporates those changes.

The fixes and removals are handled in the same `e2e/LocalSettings.php` file
that controls the settings used for the tests.

WIK-471

Co-authored-by: Vedmaka <god.vedmaka@gmail.com>

vedmaka commented on e2e/LocalSettings.php:
```diff
@@ -29,3 +30,271 @@
+	//   because it is broken, 6e4836f61b9eedf3ccc28265a3faeea31857de41 was not
+	//   backported to 1.43. Our image contains neither the `gmp` nor the `bcmath`
+	//   PHP extensions, so the ScrambleMapping class cannot be tested
+	@unlink( "$phpunitDir/unit/includes/user/TempUser/ScrambleMappingTest.php" );
```
Noting that the unlinks and the patches could probably be done in the action YAML instead

vedmaka commented on .github/workflows/docker-image.yml:
```diff
@@ -39,7 +39,91 @@ jobs:
+        run: |
+          DOCKER_BUILDKIT=1 docker compose up -d
+
+      - name: Wait for init
+        shell: bash
+        run: while ! docker compose logs web --tail 10 | grep -q '>>>>> run-maintenance-script.sh <<<<<'; do sleep 1; done
+
+      - name: Wait for container to be healthy
+        shell: bash
+        run: while ! docker inspect --format='{{json .State.Health.Status}}' $(echo $(basename $(pwd)) | tr '[:upper:]' '[:lower:]')-web-1 | grep -q 'healthy'; do sleep 1; done
```
This could make use of `docker compose up -d --wait --wait-timeout ...` instead: https://docs.docker.com/reference/cli/docker/compose/up/
